### PR TITLE
analyzer.py

### PR DESCRIPTION
**Pull request by cameron2018**
major changes:
- moved the classes from analyze.py into their own files in the models directory since that file was getting quite long and they were duplicated in there anyway.
- implemented the type eligibility and time eligibility features in analyze.py 
- created two test cases from the state database from famous people, these are used in test_analyzer.py
- added the client_builder.py which assembles a client object from parser data. this could maybe be built into the parser but i wasn't sure how to do that and i wanted a complete client object for testing the eligibilty logic.

- i also added a __dict__ method to the client and other classes which can be used to print the client object in JSON format, which i like for quick visual inspection of what the analyzer and parser are doing.

Still needs attention:
- first id like to write a bunch more in test_analyzer.py before i start refactoring the code but thats my top priority
- wrote Todo's in many places about things I'm unsure of and things that need work.
- is_charge_PCS_schedule_1 is totally untested id like to find a case in the wild to test this with
- is_crime_marijuana_list same as above
- the statute object/parser needs some work as it discards anything after the first 6 char's but that seems to work with most charges.
- the analyzer's results are just stored as properties of the client object and id like for them to return result objects like arun was doing.

I'm going to work on this more tonight and tomorrow, let me know what you think and if you see anything that i've missed or have any suggestions.

its sorta quick and dirty and I'm new to python and open source development so I would very much appreciate comments and edits.

### CODE REVIEW

**Comment by a-ilango** [comment]
Thank you @cameron2018 !

I think to start with, all the trivial commits can be squashed together with the commits that make meaningful changes. At the moment, since it's not easy to review on a per commit basis, let me look at all the changes together.

File: src/backend/expungeservice/models/statute.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        statute_string = self.statute_string
+        if len(str(statute_string))>=6: #todo: this is wrong but will kinda work for everything on our list except marijuana crimes.
+
+            statute_string = statute_string.lower() #convert to lowercase
+            statute_string = [char for char in statute_string if char not in "abcdefghijklmnopqrstuvwxyz!@#$%^&*() .,;'[]<>?:{}\""] #remove all other chars #todo: fix this it could include weird chars
+            statute_string = ''.join(statute_string)
+            statute_string = statute_string[0:6] #trim to only first 6
+
+            self.chapter = statute_string[0:3]
+            self.subchapter = statute_string[3:7]
```
Comment (a-ilango) [inline src/backend/expungeservice/models/statute.py]: We can move the parsing method to constructor. Also we should store the chapter, subchapter as ints. Integer comparison should be faster.

**Comment by a-ilango** [comment]
I have posted a few comments but haven't completely reviewed the PR yet.

File: src/backend/expungeservice/expunger/analyze.py [hunk 2]
```diff
@@ -253,55 +219,309 @@ def _have_open_case(self):
...
+                logging.info('charge is no_complaint + older than 1 year')
+                analysis = True, "Eligible under 137.225(5)(b)"
+                return analysis
+
+        elif RecordAnalyzer.is_charge_no_complaint(charge) == False:
+            logging.info('no complaint')
+            analysis = True, "Examine" #todo: i left this terse as in the flow chart. should it be "Further Analysis Needed"?
+            return analysis
+        else:
+            logging.info("this has somehow escaped the flow chart") #todo should this should throw an error?
```
[the remark this answers is not in the document]
Comment (hmarcks) [reply src/backend/expungeservice/expunger/analyze.py]: A few weeks ago I asked @michaelatcommunitylaw if No Complaint could move to Time eligibility and he agreed.

**Comment by NickSchimek** [comment]
thanks @cameronbergh Looking forward to see how this progresses.

File: src/backend/expungeservice/expunger/analyze.py [hunk 3]
```diff
@@ -253,55 +219,309 @@ def _have_open_case(self):
...
+
+        if RecordAnalyzer.is_crime_list_B(charge) == True:
+            logging.info('crime list b')
+            analysis = True, "Further Analysis Needed" #todo: 'true' doesnt seem accurate enough, maybe change this to None
+            return analysis
+
+        if RecordAnalyzer.is_crime_marijuana_list(charge) == True:  #todo: this  one is broken and will never be true
+            logging.info('crime list marijuana')
+            analysis = True, "marijuana inelgibible"
+            return analysis
```
[the remark this answers is not in the document]
Comment (cameron2018) [reply src/backend/expungeservice/expunger/analyze.py]: this list comes from the "class definitions.docx" file which was posted in the slack channel a while back. most of the items on that list include letters. and my implementation of the statute parser completely discards letters so this function is broken.

[the remark this answers is not in the document]
Comment (cameron2018) [reply src/backend/expungeservice/expunger/analyze.py]: i added "class definitions.docx" to the doc directory in git.

**Pull request closed without merging**